pkg/shardddl/pessimism/lock.go

```diff
@@ -64,13 +68,47 @@ func NewLock(id, task, owner string, ddls, sources []string) *Lock {
 // TrySync tries to sync the lock, does decrease on remain, re-entrant.
 // new upstream sources may join when the DDL lock is in syncing,
 // so we need to merge these new sources.
-func (l *Lock) TrySync(caller string, ddls, sources []string) (bool, int, error) {
+func (l *Lock) TrySync(cli *clientv3.Client, caller string, ddls, sources, latestDoneDDLs []string) (bool, int, error) {
 	l.mu.Lock()
 	defer l.mu.Unlock()
 
 	// check DDL statement first.
 	if !utils.CompareShardingDDLs(ddls, l.DDLs) {
-		return l.remain <= 0, l.remain, terror.ErrMasterShardingDDLDiff.Generate(l.DDLs, ddls)
+		curIdempotent := utils.CompareShardingDDLs(latestDoneDDLs, ddls)
+		// handle conflict
+		if !curIdempotent && !utils.CompareShardingDDLs(latestDoneDDLs, l.DDLs) {
+			return l.remain <= 0, l.remain, terror.ErrMasterShardingDDLDiff.Generate(l.DDLs, ddls)
+		}
+
+		// current ddls idempotent, skip it.
+		if curIdempotent {
```

Comment:
I still prefer that we should check `curIdempotent` before `utils.CompareShardingDDLs(ddls, l.DDLs)`

background: the lock is resolved but one non-owner is failed before saving checkpoint and resync the DDL of the resolved lock

my prefer way:
1. the failed DM-worker put the outdated DDL. if we check `curIdempotent` here, we could earlier avoid blocking that DM-worker which may caused by DDL lock

current way:
1. the failed DM-worker (worker1) put the outdated DDL, `utils.CompareShardingDDLs(ddls, l.DDLs)` is true
2. another DM-worker (worker2) put a DDL more likely a new different DDL, enter if-then branch, `curIdempotent` is false, `utils.CompareShardingDDLs(latestDoneDDLs, l.DDLs)` is true. 
    (And i think in most cases `curIdempotent` is false. And when `curIdempotent` is true, worker1 will wait more time to get skip Op)
3. in step2 we enter "other sources' ddls idempotent, skip them" branch, and here we put skip Op for worker1

Above all, worker1 may be blocked for a longer time

Reply:
And I still prefer save a binlog position through same transaction in etcd 😂 When DM-worker starting, this position has higher priority than one in downstream checkpoint